File: analyzer/indicators.py
#!/usr/bin/env python3
"""팀원2 보조 — 기술적 지표를 '코드로' 계산 (LLM은 수치 해석만).

일봉(OHLCV)에서 마감강도·거래량·MA정배열·MACD·RSI·Stochastic Slow·일목균형표를
순수 파이썬으로 산출. LLM이 직접 계산하면 환각이 나므로 반드시 여기서 수치를 만든다.

데이터: 기존 scripts/team3_price_context.fetch_daily 재사용(+ scripts/net 레이트리밋).
"""
import os
import sys

# 기존 scripts/ 재사용 (중복 구현·중복 호출 방지)
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "scripts"))
from team3_price_context import fetch_daily  # noqa: E402


def sma(xs, n):
    return sum(xs[-n:]) / n if len(xs) >= n else None


def ema_series(xs, n):
    if len(xs) < n:
        return []
    k = 2 / (n + 1)
    out = [sum(xs[:n]) / n]  # 시드 = 첫 n개 SMA
    for x in xs[n:]:
        out.append(x * k + out[-1] * (1 - k))
    return out  # 길이 = len(xs)-n+1


def macd(closes, fast=12, slow=26, sig=9):
    if len(closes) < slow + sig:
        return None
    ef, es = ema_series(closes, fast), ema_series(closes, slow)
    tail = min(len(ef), len(es))
    macd_line = [ef[-tail + i] - es[-tail + i] for i in range(tail)]
    sigs = ema_series(macd_line, sig)
    if not sigs:
        return None
    line, signal = macd_line[-1], sigs[-1]
    prev_line = macd_line[-2] if len(macd_line) >= 2 else line
    prev_sig = sigs[-2] if len(sigs) >= 2 else signal
    golden = prev_line <= prev_sig and line > signal  # 시그널 상향 돌파
    return {"macd": round(line, 2), "signal": round(signal, 2),
            "hist": round(line - signal, 2), "above_zero": line > 0,
            "golden_cross": golden, "bullish": line > signal}


def rsi(closes, n=14):
    if len(closes) < n + 1:
        return None
    gains, losses = [], []
    for i in range(1, len(closes)):
        d = closes[i] - closes[i - 1]
        gains.append(max(d, 0))
        losses.append(max(-d, 0))
    ag, al = sum(gains[:n]) / n, sum(losses[:n]) / n
    for i in range(n, len(gains)):  # Wilder 평활
        ag = (ag * (n - 1) + gains[i]) / n
        al = (al * (n - 1) + losses[i]) / n
    if al == 0:
        return {"rsi": 100.0, "zone": "과매수"}
    r = 100 - 100 / (1 + ag / al)
    zone = "과매수" if r >= 80 else "과매도" if r <= 30 else "강세" if r >= 50 else "약세"
    return {"rsi": round(r, 1), "zone": zone}


def stochastic_slow(highs, lows, closes, n=14, k=3, d=3):
    if len(closes) < n + k + d:
        return None
    fast_k = []
    for i in range(n - 1, len(closes)):
        hh = max(highs[i - n + 1:i + 1])
        ll = min(lows[i - n + 1:i + 1])
        fast_k.append(100 * (closes[i] - ll) / (hh - ll) if hh > ll else 50.0)
    slow_k = [sma(fast_k[:i + 1], k) for i in range(k - 1, len(fast_k))]
    slow_d = [sma(slow_k[:i + 1], d) for i in range(d - 1, len(slow_k))]
    if not slow_d:
        return None
    K, D = slow_k[-1], slow_d[-1]
    pk = slow_k[-2] if len(slow_k) >= 2 else K
    pd = slow_d[-2] if len(slow_d) >= 2 else D
    golden = pk <= pd and K > D
    return {"k": round(K, 1), "d": round(D, 1), "golden_cross": golden,
            "overbought": K >= 80, "bullish": K > D}


def ichimoku(highs, lows, closes):
    def midpoint(h, l, n, idx):
        if idx - n + 1 < 0:
            return None
        return (max(h[idx - n + 1:idx + 1]) + min(l[idx - n + 1:idx + 1])) / 2

    i = len(closes) - 1
    tenkan = midpoint(highs, lows, 9, i)
    kijun = midpoint(highs, lows, 26, i)
    # 오늘 위치에 그려지는 구름 = 26봉 전에 계산된 선행스팬
    j = i - 26
    if j < 0 or tenkan is None or kijun is None:
        return {"available": False}
    t26, k26 = midpoint(highs, lows, 9, j), midpoint(highs, lows, 26, j)
    spanA = (t26 + k26) / 2 if t26 and k26 else None
    spanB = midpoint(highs, lows, 52, j)
    if spanA is None or spanB is None:
        return {"available": False}
    cloud_top, cloud_bot = max(spanA, spanB), min(spanA, spanB)
    close = closes[-1]
    return {"available": True,
            "above_cloud": close > cloud_top,
            "in_cloud": cloud_bot <= close <= cloud_top,
            "tenkan_gt_kijun": tenkan > kijun,
            "tenkan": round(tenkan, 1), "kijun": round(kijun, 1),
            "cloud_top": round(cloud_top, 1), "cloud_bot": round(cloud_bot, 1)}


def compute_indicators(code, name=None, days=200):
    """종목코드 → 기술적 지표 묶음(수치 + 강세 플래그)."""
    name = name or code
    try:
        d = fetch_daily(code, days=days)
    except Exception as e:
        return {"code": code, "name": name, "error": str(e)}
    if len(d) < 35:
        return {"code": code, "name": name, "error": f"데이터부족({len(d)})"}

    closes = [x["close"] for x in d]
    highs = [x["high"] for x in d]
    lows = [x["low"] for x in d]
    vols = [x["volume"] for x in d]
    last, hi, lo = closes[-1], highs[-1], lows[-1]

    ma5, ma20, ma60 = sma(closes, 5), sma(closes, 20), sma(closes, 60)
    aligned = bool(ma5 and ma20 and ma60 and ma5 > ma20 > ma60)  # 정배열
    close_strength = round((last - lo) / (hi - lo), 2) if hi > lo else None  # 마감강도(1=고가마감)
    vol_avg20 = sma(vols, 20)
    vol_ratio = round(vols[-1] / vol_avg20, 2) if vol_avg20 else None

    return {
        "code": code, "name": name, "last_close": last,
        "ma_aligned": aligned, "ma5": ma5, "ma20": ma20, "ma60": ma60,
        "close_strength": close_strength,  # 0~1, 0.8+ = 강세 마감
        "volume_vs_20d": vol_ratio,
        "macd": macd(closes),
        "rsi": rsi(closes),
        "stochastic_slow": stochastic_slow(highs, lows, closes),
        "ichimoku": ichimoku(highs, lows, closes),
    }


if __name__ == "__main__":
    import json
    code = sys.argv[1] if len(sys.argv) > 1 else "018880"
    name = sys.argv[2] if len(sys.argv) > 2 else code
    print(json.dumps(compute_indicators(code, name), ensure_ascii=False, indent=2))

File: web/lib/stock/indicators.ts
// 기술적 지표 — analyzer/indicators.py 1:1 포팅 (순수 함수, 환각 없는 코드 계산).
// 입력은 일봉 배열(과거→최신 정렬), 최소 35봉 미만이면 null.

import type { Candle, TechnicalSection } from "@/types/stock";

export function sma(xs: number[], n: number): number | null {
  if (xs.length < n) return null;
  let s = 0;
  for (let i = xs.length - n; i < xs.length; i++) s += xs[i];
  return s / n;
}

/** EMA 시리즈 — 시드는 첫 n개 SMA. 길이 = len-n+1. */
function emaSeries(xs: number[], n: number): number[] {
  if (xs.length < n) return [];
  const k = 2 / (n + 1);
  let seed = 0;
  for (let i = 0; i < n; i++) seed += xs[i];
  const out = [seed / n];
  for (let i = n; i < xs.length; i++) {
    out.push(xs[i] * k + out[out.length - 1] * (1 - k));
  }
  return out;
}

export function macd(
  closes: number[],
  fast = 12,
  slow = 26,
  sig = 9
): TechnicalSection["macd"] {
  if (closes.length < slow + sig) return null;
  const ef = emaSeries(closes, fast);
  const es = emaSeries(closes, slow);
  const tail = Math.min(ef.length, es.length);
  const line: number[] = [];
  for (let i = 0; i < tail; i++) {
    line.push(ef[ef.length - tail + i] - es[es.length - tail + i]);
  }
  const sigs = emaSeries(line, sig);
  if (sigs.length === 0) return null;
  const m = line[line.length - 1];
  const s = sigs[sigs.length - 1];
  const pm = line.length >= 2 ? line[line.length - 2] : m;
  const ps = sigs.length >= 2 ? sigs[sigs.length - 2] : s;
  return {
    macd: round2(m),
    signal: round2(s),
    hist: round2(m - s),
    aboveZero: m > 0,
    goldenCross: pm <= ps && m > s, // 시그널 상향 돌파
    bullish: m > s,
  };
}

export function rsi(closes: number[], n = 14): TechnicalSection["rsi"] {
  if (closes.length < n + 1) return null;
  const gains: number[] = [];
  const losses: number[] = [];
  for (let i = 1; i < closes.length; i++) {
    const d = closes[i] - closes[i - 1];
    gains.push(Math.max(d, 0));
    losses.push(Math.max(-d, 0));
  }
  let ag = 0;
  let al = 0;
  for (let i = 0; i < n; i++) {
    ag += gains[i];
    al += losses[i];
  }
  ag /= n;
  al /= n;
  for (let i = n; i < gains.length; i++) {
    // Wilder 평활
    ag = (ag * (n - 1) + gains[i]) / n;
    al = (al * (n - 1) + losses[i]) / n;
  }
  const r = al === 0 ? 100 : 100 - 100 / (1 + ag / al);
  const value = Math.round(r * 10) / 10;
  const zone = r >= 80 ? "과매수" : r <= 30 ? "과매도" : r >= 50 ? "강세" : "약세";
  return { value, zone };
}

export function stochasticSlow(
  highs: number[],
  lows: number[],
  closes: number[],
  n = 14,
  k = 3,
  d = 3
): TechnicalSection["stochastic"] {
  if (closes.length < n + k + d) return null;
  const fastK: number[] = [];
  for (let i = n - 1; i < closes.length; i++) {
    let hh = -Infinity;
    let ll = Infinity;
    for (let j = i - n + 1; j <= i; j++) {
      if (highs[j] > hh) hh = highs[j];
      if (lows[j] < ll) ll = lows[j];
    }
    fastK.push(hh > ll ? (100 * (closes[i] - ll)) / (hh - ll) : 50.0);
  }
  const trailingAvg = (xs: number[], end: number, m: number) => {
    let s = 0;
    for (let j = end - m + 1; j <= end; j++) s += xs[j];
    return s / m;
  };
  const slowK: number[] = [];
  for (let i = k - 1; i < fastK.length; i++) slowK.push(trailingAvg(fastK, i, k));
  const slowD: number[] = [];
  for (let i = d - 1; i < slowK.length; i++) slowD.push(trailingAvg(slowK, i, d));
  if (slowD.length === 0) return null;
  const K = slowK[slowK.length - 1];
  const D = slowD[slowD.length - 1];
  const pk = slowK.length >= 2 ? slowK[slowK.length - 2] : K;
  const pd = slowD.length >= 2 ? slowD[slowD.length - 2] : D;
  return {
    k: Math.round(K * 10) / 10,
    d: Math.round(D * 10) / 10,
    goldenCross: pk <= pd && K > D,
    overbought: K >= 80,
    bullish: K > D,
  };
}

export function ichimoku(
  highs: number[],
  lows: number[],
  closes: number[]
): TechnicalSection["ichimoku"] {
  const midpoint = (n: number, idx: number): number | null => {
    if (idx - n + 1 < 0) return null;
    let hh = -Infinity;
    let ll = Infinity;
    for (let j = idx - n + 1; j <= idx; j++) {
      if (highs[j] > hh) hh = highs[j];
      if (lows[j] < ll) ll = lows[j];
    }
    return (hh + ll) / 2;
  };
  const i = closes.length - 1;
  const tenkan = midpoint(9, i);
  const kijun = midpoint(26, i);
  // 오늘 위치에 그려지는 구름 = 26봉 전에 계산된 선행스팬
  const j = i - 26;
  if (j < 0 || tenkan === null || kijun === null) return { available: false };
  const t26 = midpoint(9, j);
  const k26 = midpoint(26, j);
  const spanA = t26 !== null && k26 !== null ? (t26 + k26) / 2 : null;
  const spanB = midpoint(52, j);
  if (spanA === null || spanB === null) return { available: false };
  const cloudTop = Math.max(spanA, spanB);
  const cloudBot = Math.min(spanA, spanB);
  const close = closes[closes.length - 1];
  return {
    available: true,
    aboveCloud: close > cloudTop,
    inCloud: cloudBot <= close && close <= cloudTop,
    tenkanGtKijun: tenkan > kijun,
    tenkan: round1(tenkan),
    kijun: round1(kijun),
    cloudTop: round1(cloudTop),
    cloudBot: round1(cloudBot),
  };
}

/** 일봉 → 지표 묶음. 35봉 미만(신규상장 등)이면 null. */
export function computeIndicators(candles: Candle[]): TechnicalSection | null {
  if (candles.length < 35) return null;
  const closes = candles.map((c) => c.close);
  const highs = candles.map((c) => c.high);
  const lows = candles.map((c) => c.low);
  const vols = candles.map((c) => c.volume);
  const last = candles[candles.length - 1];

  const ma5 = sma(closes, 5);
  const ma20 = sma(closes, 20);
  const ma60 = sma(closes, 60);
  const volAvg20 = sma(vols, 20);

  return {
    ma5: ma5 !== null ? Math.round(ma5) : null,
    ma20: ma20 !== null ? Math.round(ma20) : null,
    ma60: ma60 !== null ? Math.round(ma60) : null,
    maAligned: ma5 !== null && ma20 !== null && ma60 !== null && ma5 > ma20 && ma20 > ma60,
    closeStrength:
      last.high > last.low
        ? Math.round(((last.close - last.low) / (last.high - last.low)) * 100) / 100
        : null,
    volumeVs20d:
      volAvg20 && volAvg20 > 0 ? Math.round((vols[vols.length - 1] / volAvg20) * 100) / 100 : null,
    macd: macd(closes),
    rsi: rsi(closes),
    stochastic: stochasticSlow(highs, lows, closes),
    ichimoku: ichimoku(highs, lows, closes),
  };
}

const round1 = (x: number) => Math.round(x * 10) / 10;
const round2 = (x: number) => Math.round(x * 100) / 100;
